storage/imxrt-flash/flashsrv.c

```diff
@@ -12,78 +12,323 @@
  */
 
 #include <errno.h>
+#include <stdlib.h>
+#include <string.h>
 #include <unistd.h>
 #include <sys/msg.h>
 #include <sys/file.h>
+#include <sys/threads.h>
+#include <posix/utils.h>
 
-#include "posix/utils.h"
+#include <meterfs.h>
+#include <ptable.h>
 
 #include "flashdrv.h"
 #include "flashsrv.h"
 
-#define FLASHES_NO 2
 
 #define LOG_ERROR(str, ...) do { fprintf(stderr, __FILE__  ":%d error: " str "\n", __LINE__, ##__VA_ARGS__); } while (0)
 #define TRACE(str, ...) do { if (0) fprintf(stderr, __FILE__  ":%d trace: " str "\n", __LINE__, ##__VA_ARGS__); } while (0)
 
+#define THREAD_STACKSZ               512
+#define THREAD_PRIORITY              4
+
+#define FLASH_MEMORIES_NO            2
+
 
 enum { flash_memory_unactive = 0, flash_memory_active = 0xff };
 
 
-typedef struct _flash_memory_t {
-	flash_context_t context;
+typedef struct flashsrv_partition {
+	ptable_partition_t *pHeader;
+
 	oid_t oid;
-	uint8_t status;
+	uint8_t fID;
+	uint8_t pStatus;
+
+	void *fsCtx;
+} flashsrv_partition_t;
+
+
+typedef struct flash_memory {
+	flash_context_t ctx;
+	flashsrv_partition_t *parts;
+
+	uint32_t pCnt;
+	uint8_t fStatus;
+	uint8_t currPart;
+	uint8_t rawActive;
+
+	oid_t fOid;
+	uint32_t rawPort;
+
+	handle_t lock;
 } flash_memory_t;
 
 
 struct {
-	flash_memory_t flash_memories[FLASHES_NO];
-	uint32_t flexspi_adresses[FLASHES_NO];
+	flash_memory_t flash_memories[FLASH_MEMORIES_NO];
+	uint32_t flexspi_adresses[FLASH_MEMORIES_NO];
 } flashsrv_common;
 
 
-static flash_memory_t *flashsrv_findFlash(oid_t oid)
+
+/* Flash functions */
+
+static int flashsrv_bufferedPagesWrite(uint8_t fID, size_t offset, const char *data, size_t size)
+{
+	flash_memory_t *flash_memory;
+
+	if (FLASH_MEMORIES_NO <= fID)
+		return -EINVAL;
```

Comment:
Maybe we should perform this check only directly in the interface function (i.e. message receiving and parsing) to avoid unnecessary defensiveness